designs/2024-multithread-linting/README.md

```diff
@@ -0,0 +1,780 @@
+
+        // Calculate effective concurrency.
+        const concurrency = Math.min(eslintOptions.concurrency, filePaths.length);
+        debug(`Running ${concurrency} worker thread(s).`);
+
+        // Start worker threads.
+        const abortController = new AbortController();
+        const fileIndexCounter = new Int32Array(new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT));
+        const workerPath = require.resolve("./worker.js");
+        const workerOptions = {
+            workerData: {
+                filePaths,
+                fileIndexCounter,
+                eslintOptions
+            },
+        };
+        const workerExecutor = function (resolve, reject) {
+            const worker = new Worker(workerPath, workerOptions);
+
+            // A message is only received once when the thread exist successfully.
+            worker.once(
+                "message",
+                indexedResults => {
+                    for (const result of indexedResults) {
+                        const { index } = result;
+                        delete result.index;
+                        results[index] = result;
+                    }
+                    resolve();
+                }
+            );
+            worker.once(
+                "error",
+                error => {
+                    abortController.abort(error);
+                    reject(error);
+                }
+            );
+            abortController.signal.addEventListener("abort", () => worker.terminate());
+        };
+        const promises = new Array(concurrency);
+
+        for (let index = 0; index < concurrency; ++index) {
+            promises[index] = new Promise(workerExecutor);
+        }
+
+        // Wait until all results have been collected.
+        await Promise.all(promises);
+
+        // Persist the cache to disk.
+        if (lintResultCache) {
+            results.forEach((result, index) {
+                if (result) {
+                    const filePath = filePaths[index];
+                    const configs = configLoader.getCachedConfigArrayForFile(filePath);
+                    const config = configs.getConfig(filePath);
+
+                    lintResultCache.setCachedLintResults(filePath, config, result);
+                }
+            );
+            lintResultCache.reconcile();
+        }
+```
+
+### Worker threads
+
+In JavaScript, threads have separate execution contexts and access separate memory.
+This allows diferent threads in the same process to efficiently run on different cores without locking each other.
+The thread's entry point is a module specifier (path or URL) that is loaded as soon as the thread starts.
+
+For the purpose of multithread linting, all linting threads are created by the controlling thread at the same time before the linting process starts.
+Although the controlling thread will be typically the main thread of the process, this is not a general requirement.
```

Comment:
Mentioned above, but noting here as well: I believe the real worker count should be informed by the size of the enumerated file list. worker threads are much lighter than full processes, but I believe they still have some cost that means default behavior on small projects will be slower than optimal. This could be tested once an implementation exists to know for sure, though

Reply:
Indeed yes, there is no point in having more worker threads than there are files to be linted. In `"auto"` concurreny mode, the number of worker threads could be limited to one worker thread each 50 files, for example. I would like to come up with a proto-implementation to determine the number of effective threads, even if the logic could change anytime.

Reply:
I've update the RFC to show how the number of files could affect the effective number of worker threads in https://github.com/eslint/rfcs/blob/multithread-linting/designs/2024-multithread-linting/README.md#number-of-worker-threads. Thanks for the suggestion.